src/controllers/controllerengine.cpp

```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
     }
 }
 
-/* -------- ------------------------------------------------------
-Purpose: Resolves a function name to a QScriptValue including
-            OBJECT.Function calls
-Input:   -
-Output:  -
--------- ------------------------------------------------------ */
-QScriptValue ControllerEngine::resolveFunction(const QString& function) const {
-    QHash<QString, QScriptValue>::const_iterator i =
-            m_scriptValueCache.find(function);
-    if (i != m_scriptValueCache.end()) {
-        return i.value();
-    }
+/* ------------------------------------------------------------------
+Purpose: Turn a snippet of JS into a QScriptValue function.
+         Wrapping it in an anonymous function allows any JS that
+         evaluates to a function to be used in MIDI mapping XML files
+         and ensures the function is executed with the correct
+         'this' object.
+Input:   QString snippet of JS that evaluates to a function,
+         int number of arguments that the function takes
+Output:  QScriptValue of JS snippet wrapped in an anonymous function
+------------------------------------------------------------------- */
+QScriptValue ControllerEngine::wrapFunctionCode(const QString& codeSnippet,
+                                                int numberOfArgs) {
+    QScriptValue wrappedFunction;
 
-    QScriptValue object = m_pEngine->globalObject();
-    QStringList parts = function.split(".");
+    QHash<QString, QScriptValue>::const_iterator i =
+            m_scriptWrappedFunctionCache.find(codeSnippet);
 
-    for (int i = 0; i < parts.size(); i++) {
-        object = object.property(parts.at(i));
-        if (!object.isValid()) {
-            break;
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
+        if (!syntaxIsValid(codeSnippet)) {
+            wrappedFunction = m_pEngine->evaluate(
+                "(function () { print('Syntax error in: "
```

Comment:
If we print "Syntax error" from the handler we produce here that means that it will get printed on every button press (not just the first) after we've already delivered a popup dialog to the user telling them that stuff is going wrong with their scripts.

If this is a script in the wild that's misbehaving this will result in spamming the user's log (which they probably aren't looking at it if they aren't technical). If this were say a jog wheel handler, that means we'd print an error message for every tick of the jog wheel. Printing too many error messages can actually slow down mixxx via priority inversion since any other thread that prints has to fight for a mutex to write to stdout.